File: gateway/internal/handlers/creation_draft_edit.go
package handlers

import (
	"encoding/json"
	"errors"
	"reflect"

	"github.com/aan/agent-assistant-gateway/internal/bridge"
	"github.com/aan/agent-assistant-gateway/internal/models"
)

// Edit sources belong to the same node's draft history, not its identity inputs.
// Keep the original reference graph for independent acceptance review.
func validateEditSources(doc creativeDocument, plan bridge.CreativePlan, req bridge.CreationPlanningRequest) error {
	locked := map[string]bool{}
	for _, id := range req.LockedNodeIDs {
		locked[id] = true
	}
	for _, node := range plan.Nodes {
		if node.EditSourceAssetID == "" {
			continue
		}
		if req.Repair != nil && req.Repair.NodeID == node.ID && exhaustedEdits(req) {
			return errors.New("连续三次编辑仍未通过同一要求，请清空编辑来源并重新规划生成")
		}
		before, exists := creativeNode(doc, node.ID)
		if !exists || before.Kind != "image" || node.Kind != "image" {
			return errors.New("编辑草稿必须来自当前已有图片节点")
		}
		if creationJSON(before.References) != creationJSON(node.References) || !reflect.DeepEqual(before.DependsOn, node.DependsOn) || before.AspectRatio != node.AspectRatio {
			return errors.New("局部编辑不能同时改变原参考关系或画幅，请重新生成")
		}
		if !sameDraftAncestors(doc, creativeDocument{Plan: plan}, node, false) {
			return errors.New("局部编辑不能同时更换上游内容，请清空编辑来源后重新生成")
		}
		if before.EditSourceAssetID == node.EditSourceAssetID {
			continue
		}
		if locked[node.ID] {
			return errors.New("不能编辑已确认节点的草稿")
		}
		allowed := false
		if req.Repair != nil {
			if req.AutomaticMode && req.Repair.NodeID == node.ID {
				for _, id := range req.Repair.CandidateIDs {
					allowed = allowed || id == node.EditSourceAssetID
				}
			}
		} else {
			state := doc.States[node.ID]
			allowed = state.SelectedAssetID == node.EditSourceAssetID
			for _, id := range state.Candidates {
				allowed = allowed || id == node.EditSourceAssetID
			}
		}
		if !allowed {
			return errors.New("待修草稿必须是当前节点自己的候选")
		}
	}
	return nil
}

// A draft embeds its ancestor context even though the editor only receives one
// image. Check that context independently of the current graph's approval state.
func sameDraftAncestors(before, after creativeDocument, node bridge.CreativeNode, checkState bool) bool {
	pending := append([]string{}, node.DependsOn...)
	seen := map[string]bool{}
	for len(pending) > 0 {
		id := pending[len(pending)-1]
		pending = pending[:len(pending)-1]
		if seen[id] {
			continue
		}
		seen[id] = true
		old, oldOK := creativeNode(before, id)
		current, currentOK := creativeNode(after, id)
		old.RevisionSuggestions, current.RevisionSuggestions = nil, nil
		if !oldOK || !currentOK || !creativeNodesEqual(old, current) {
			return false
		}
		if checkState {
			a, b := before.States[id], after.States[id]
			if a.Revision != b.Revision || a.SelectedAssetID != b.SelectedAssetID {
				return false
			}
		}
		pending = append(pending, current.DependsOn...)
	}
	return true
}

func (h *CreationHandler) validateDraftContext(row models.CreationProject, doc creativeDocument, node bridge.CreativeNode) error {
	if node.EditSourceAssetID == "" {
		return nil
	}
	if err := h.validateDraftProvenance(row, node); err != nil {
		return err
	}
	var run models.CreationRun
	if err := h.db.Table("creation_runs AS r").Select("r.*").Joins("JOIN creation_assets AS a ON a.run_id = r.id AND a.user_id = r.user_id").
		Where("a.id = ? AND a.user_id = ? AND r.project_id = ? AND r.project_node_id = ?", node.EditSourceAssetID, row.UserID, row.ID, node.ID).Take(&run).Error; err != nil {
		return errors.New("无法核对待修草稿的原始上下文")
	}
	var source creativeDocument
	if json.Unmarshal([]byte(run.Snapshot), &source) != nil {
		return errors.New("待修草稿缺少原始上下文，请重新生成")
	}
	original, exists := creativeNode(source, node.ID)
	if !exists || creationJSON(original.References) != creationJSON(node.References) ||
		!reflect.DeepEqual(original.DependsOn, node.DependsOn) || original.AspectRatio != node.AspectRatio ||
		!sameDraftAncestors(source, doc, node, true) {
		return errors.New("待修草稿的上游内容或参考已变化，请清空编辑来源后重新规划生成")
	}
	return nil
}

func (h *CreationHandler) validateDraftProvenance(row models.CreationProject, node bridge.CreativeNode) error {
	if node.EditSourceAssetID == "" {
		return nil
	}
	var count int64
	err := h.db.Table("creation_assets AS a").Joins("JOIN creation_runs AS r ON r.id = a.run_id AND r.user_id = a.user_id").
		Where("a.id = ? AND a.user_id = ? AND r.project_id = ? AND r.project_node_id = ?", node.EditSourceAssetID, row.UserID, row.ID, node.ID).Count(&count).Error
	if err != nil || count != 1 {
		return errors.New("待修草稿的项目或节点来源不匹配")
	}
	return nil
}

File: agent/aigc/creation_draft_edit.py
"""Working drafts retain provenance without becoming approved references."""

DRAFT_EDIT_GUIDANCE = '''
图片返工有两种操作：重新生成，或编辑当前节点的待修草稿。候选已基本符合身份、环境与构图、仅有可局部修正的问题（如人物占比）时，优先考虑局部编辑，避免整张重画丢失正确内容。
局部编辑使用edit_source_asset_id填写本轮repair.candidate_ids中的一张真实候选；prompt仅写基于Picture 1的具体修改、必须保留的内容和原目标。references、depends_on、画幅保持不变，character_style与asset_id为空。系统只向编辑器传该草稿像素，不重新注入原人设或场景；原references继续作为复审依据。无法在现有草稿中可靠修正的身份混淆、整体场景错误应重新生成，明确edit_source_asset_id=""并恢复完整生成prompt。不要承诺像素完全不变，不放宽冻结要求。新图是未审阅候选，不能直接批准，也不能把旧失败图放进其他节点的asset_id或references。
'''


def validate_edit_sources(plan, request):
    from agent.aigc.creation_planning import CreativePlan
    from agent.aigc.creation_output import omit_null_fields
    if not any(n.edit_source_asset_id for n in plan.nodes):
        return
    previous = {n.id: n for n in CreativePlan.model_validate(omit_null_fields(request.current_plan)).nodes}
    proposed = {n.id: n for n in plan.nodes}
    for node in plan.nodes:
        source = node.edit_source_asset_id
        if not source:
            continue
        before = previous.get(node.id)
        if not before or before.kind != 'image':
            raise ValueError('编辑草稿必须来自当前已有图片节点：' + node.id)
        for field in ('references', 'depends_on', 'aspect_ratio'):
            if getattr(before, field) != getattr(node, field):
                raise ValueError('局部编辑不能同时改变原参考关系或画幅，请清空edit_source_asset_id后重新生成：' + node.id)
        pending, seen = list(node.depends_on), set()
        while pending:
            dep = pending.pop()
            if dep in seen:
                continue
            seen.add(dep)
            old, new = previous.get(dep), proposed.get(dep)
            if not old or not new or old.model_dump(exclude={'revision_suggestions'}) != new.model_dump(exclude={'revision_suggestions'}):
                raise ValueError('局部编辑不能同时更换上游内容，请清空edit_source_asset_id后重新生成：' + node.id)
            pending.extend(new.depends_on)
        if source == before.edit_source_asset_id:
            continue
        if node.id in request.locked_node_ids:
            raise ValueError('不能编辑已确认节点的草稿：' + node.id)
        if request.repair:
            allowed = request.repair.candidate_ids if request.automatic_mode and request.repair.node_id == node.id else []
        else:
            state = request.node_context.get(node.id, {})
            allowed = [state.get('selected_asset_id', '')] + state.get('candidate_asset_ids', [])
        if source not in allowed:
            raise ValueError('待修草稿必须是当前节点自己的候选：' + node.id)
